### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by dv343**
This is a complete rewrite of the repo, it tries to be easier to understand and maintain, and more importantly it fixes https://github.com/CartoDB/tangram-cartocss/issues/63

I expect better performance too, since the output YAML only uses JS functions when there is a dependency on features properties, and it output literals otherwise.

Filtering is done using Tangram's `filter` instead of by modifying the style.

Tests cases cover many cases and most corner-cases, especially the ones that are different from the implementation point of view.

However, test cases for filtering are missing right now.

This supersedes #64 and #65.

### CODE REVIEW

**Comment by rochoa** [comment]
How does this affect the work in #65?

**Comment by dv343** [comment]
The translation of blending mentioned in https://github.com/CartoDB/tangram-cartocss/pull/65 is done, Tangram's problem is still there, and related decisions (should we mark as supported something that partially works?) are delegated to tangram-reference. Whatever we decide, this should work, the necessary translation is done, and we can mark as unsupported blending modes on tangram-reference.

**Comment by dv343** [comment]
I've just added a few test cases for filtering and it works OK.
There are probably many edge cases untested, but still, there are more tests than before.

File: doc/about_rendering.md [hunk 1]
```diff
@@ -1,66 +0,0 @@
-# About Rendering
```
Comment (rochoa) [inline doc/about_rendering.md]: Shouldn't we keep any of this?

File: index.html [hunk 2]
```diff
@@ -1,35 +0,0 @@
-<!DOCTYPE html>
```
Comment (rochoa) [inline index.html]: Can we keep any kind of demo?

File: package.json [hunk 3]
```diff
@@ -1,13 +1,25 @@
 {
   "name": "tangram-cartocss",
-  "version": "0.6.0",
+  "version": "1.0.0",
```
Comment (rochoa) [inline package.json]: I would keep version <1.0.0 as we might want to change the interface.

File: package.json [hunk 4]
```diff
@@ -1,13 +1,25 @@
...
   "description": "Transform cartocss into a draw tangram object.",
+  "main": "src/index.js",
   "repository": {
     "type": "git",
     "url": "git://github.com/cartodb/tangram-cartocss.git"
   },
-  "main": "dist/tangram-cartocss.debug.js",
-  "homepage": "https://github.com/cartodb/tangram-cartocss",
+  "dependencies": {
+    "carto": "github:cartodb/carto#dasharrayJSsupport",
```
Comment (rochoa) [inline package.json]: We need to merge that branch before merging this one.

File: package.json [hunk 5]
```diff
@@ -1,13 +1,25 @@
...
+    "tangram-reference": "github:cartodb/tangram-reference#inconsistency-fix"
```
Comment (rochoa) [inline package.json]: Same here.

File: src/basic/text.js [hunk 6]
```diff
@@ -1,77 +0,0 @@
-/**
```
Comment (rochoa) [inline src/basic/text.js]: I'm a bit curious. Did we have support for texts?

File: src/color.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    function hexToInt(hex) {
+        return parseInt('0x' + hex);
+    }
+    var result = {};
+    if (color[0] === '#') {
+        result.r = hexToInt(color.substr(1, 2));
+        result.g = hexToInt(color.substr(3, 2));
+        result.b = hexToInt(color.substr(5, 2));
+        result.a = 1;
+    } else if (color.startsWith('rgb(')) {
```
Comment (rochoa) [inline src/color.js]: Will we need a polyfill to use `startsWith`?

File: src/index.js [hunk 8]
```diff
@@ -0,0 +1,277 @@
+const tangramReference = require('tangram-reference').load();
+const color = require('./color.js');
+
+//getReferenceIndexedByCSS returns a reference based on ref indexed by the CSS property name instead of the reference original property name
+//it does a shallow copy of each property, therefore, it's possible to change the returned reference by changing the original one, don't do it
```
Comment (rochoa) [inline src/index.js]: Wouldn't be better to avoid shallow copy and do a deep copy?

File: src/index.js [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+            newRef[ref.symbolizers[symb][property].css] = ref.symbolizers[symb][property];
+        }
+    }
+    return newRef;
+}
+const referenceCSS = getReferenceIndexedByCSS(tangramReference);
+
+
+function wrapFn(functionBody) {
+    return 'function (){var data=feature; var ctx={zoom: $zoom};' + functionBody + '}';
```
Comment (rochoa) [inline src/index.js]: I would use template literal as we have done in the rest of the code.

File: src/index.js [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+        opacity = yamlDrawGroup._hidden['opacity:fill'];
+    } else {
+        opacity = yamlDrawGroup._hidden['opacity:outline'];
+    }
+    opacity = yamlDrawGroup._hidden['opacity:general'] || opacity;
+    return opacity;
+}
+
+function isNumeric(n) {
+    return !isNaN(parseFloat(n)) && Number.isFinite(n);
```
Comment (rochoa) [inline src/index.js]: This is exactly the same as doing `Number.isFinite(n)` alone, isn't it?

File: src/index.js [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+    const ctx = { zoom: 10 };
+    // jshint ignore:end
+    var _value = null;
+    shaderValue.js.forEach(function (code) {
+        eval(code);
+    });
+    return _value;
+}
+
+//Returns the final opacity override selecting between fill-opacity and outline-opacity.
```
Comment (rochoa) [inline src/index.js]: I would consider using [JSDoc](http://usejsdoc.org/). What do you think?

File: src/index.js [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+        processStyle(yaml, drawGroupName, layerOrder);
+    }
+    return yaml;
+}
+module.exports.layerToYAML = layerToYAML;
+module.exports.carto2Draw = layerToYAML;
+
+
+/*
+//Usage example
```
Comment (rochoa) [inline src/index.js]: I would keep this in the readme.

File: src/index.js [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+        const drawGroupName = `drawGroup${layerOrder}`;
+        processPoints(yaml, layer, drawGroupName);
+        processLines(yaml, layer, drawGroupName);
+        processPolys(yaml, layer, drawGroupName);
+
+        processStyle(yaml, drawGroupName, layerOrder);
+    }
+    return yaml;
+}
+module.exports.layerToYAML = layerToYAML;
```
Comment (rochoa) [inline src/index.js]: I don't know if this is the best naming, as it doesn't return a YAML, but a js/json.

The same applies to all the functions (`processPoints `, ...) receiving it as a param.

File: doc/about_rendering.md (hunk 1 above)
Comment (dv343) [reply doc/about_rendering.md]: The first limitation (dynamic marker-width) is resolved with this (it's #64 ).

Regarding the second limitation (allow-overlap)... I think there is no point in saying this here. The limitation is still valid, but the problem resides in Tangram, and if tangram-reference it's ok, then the CartoCSSRenderer will throw an appropriate exception to fallback to raster.
There are many limitations, not only this one, I think this could be said in tangram-reference, but not here, particularly because this is Tangram's fault.
However, we could say here that texts aren't supported, since this is tangram-css fault, not Tangram's.

File: index.html (hunk 2 above)
Comment (dv343) [reply index.html]: I don't see the point since tangram-css is relatively low-level, I think that the demo should be at tangram.cartodb. There is an usage example in the Readme that could be used instead.

File: package.json (hunk 3 above)
Comment (dv343) [reply package.json]: Yes, "1.0.0" was the default value, do you think that "0.7.0" is OK?

File: package.json (hunk 4 above)
Comment (dv343) [reply package.json]: Yes

File: package.json (hunk 5 above)
Comment (dv343) [reply package.json]: Yes

File: src/basic/text.js (hunk 6 above)
Comment (dv343) [reply src/basic/text.js]: We saw the other day that it didn't work, maybe it worked in some cases, maybe it was implemented but never tested.
I wouldn't hold this PR for the lack of text support since I think it was broken already and now it's just not implemented.

File: src/color.js (hunk 7 above)
Comment (dv343) [reply src/color.js]: It's a good question, I could look at `startsWith` docs, but I've worked without limiting myself to a particular ES version. Is there a way to automatically detect which ES version works? Do we want to configure babel?

File: doc/about_rendering.md (hunk 1 above)
Comment (rochoa) [reply doc/about_rendering.md]: Fair enough. I would keep the information about text/fonts.

File: index.html (hunk 2 above)
Comment (rochoa) [reply index.html]: 👍

File: package.json (hunk 3 above)
Comment (rochoa) [reply package.json]: Yes. 😈

File: src/basic/text.js (hunk 6 above)
Comment (rochoa) [reply src/basic/text.js]: Agree. I wanted to keep these comments here for future reference. Also, we will have this information in the `about_rendering.md` document.

File: src/index.js (hunk 8 above)
Comment (dv343) [reply src/index.js]: It will be certainly safer (if someone modifies it, bad things could happen) but slower. I left it like this because modifying a reference at runtime makes no sense to me, and if we need to modify it, the modification will be done in that same file because the reference is loaded there.
I would leave it like this.

File: src/index.js (hunk 9 above)
Comment (dv343) [reply src/index.js]: Yes

File: src/index.js (hunk 8 above)
Comment (rochoa) [reply src/index.js]: 👌

File: src/index.js (hunk 11 above)
Comment (dv343) [reply src/index.js]: I don't have experience with it, but I could use it. However, should we hold this PR for this?

File: src/index.js (hunk 10 above)
Comment (dv343) [reply src/index.js]: Good point! Yes, is redundant, and there is another problem, you mentioned before the polyfill for `startsWith`, but this could be a problematic function too. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isFinite it isn't supported on IE

File: src/index.js (hunk 11 above)
Comment (rochoa) [reply src/index.js]: No, no need to hold it. I just wanted to raise some awareness :).

File: src/index.js (hunk 13 above)
Comment (dv343) [reply src/index.js]: That's true, although it is the object that is used to generate the YAML, and it doesn't suffer more transformations afterwards. Do you think of a better name?

File: src/index.js (hunk 12 above)
Comment (dv343) [reply src/index.js]: A simpler version of this it's already on the readme, should I remove this one?

File: src/index.js (hunk 13 above)
Comment (dv343) [reply src/index.js]: Renamed to "scene"
